ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java

```diff
@@ -261,21 +288,32 @@
   @Override
   public boolean hasSucceeded() {
     boolean hasSucceeded;
-    if (routerOperation == RouterOperation.PutOperation && routerConfig.routerPutUseDynamicSuccessTarget) {
-      hasSucceeded =
-          succeededCount >= Math.max(totalReplicaCount - disabledCount - 1, routerConfig.routerPutSuccessTarget);
+    if (routerOperation == RouterOperation.PutOperation && routerConfig.routerPutUseDynamicSuccessTarget
+        && inFlightReplicaType == ReplicaType.DISK_BACKED) {
+      // this logic only applies to disk replicas where the quorum can change during replica movement
+      int dynamicSuccessTarget = Math.max(totalReplicaCount - disabledCount - 1, routerConfig.routerPutSuccessTarget);
+      hasSucceeded = diskReplicaSuccessCount >= dynamicSuccessTarget;
     } else {
-      hasSucceeded = succeededCount >= successTarget;
+      hasSucceeded = diskReplicaSuccessCount >= diskSuccessTarget || cloudReplicaSuccessCount >= cloudSuccessTarget;
     }
     return hasSucceeded;
   }
 
   @Override
   public boolean hasFailedOnNotFound() {
-    return (originatingDcNotFoundFailureThreshold > 0
-        && originatingDcNotFoundCount >= originatingDcNotFoundFailureThreshold) || (crossColoEnabled
-        && diskDownCount + totalNotFoundCount > totalReplicaCount - successTarget);
+    if (routerOperation == RouterOperation.PutOperation) {
+      return false;
+    }
+    if (originatingDcNotFoundFailureThreshold > 0
+        && originatingDcNotFoundCount >= originatingDcNotFoundFailureThreshold) {
+      return true;
+    }
     // To account for GET operation, the threshold should be  >= totalReplicaCount - (success target - 1)
+    // Right now, this only applies for disk replica only partitions and may not be completely accurate if there are
+    // failures responses other than not found.
+    // TODO support cloud replicas in this condition, also account for failures other than not found
```

Comment:
Is there still a TODO here?

Reply:
Yes that is still a todo. I will make github issues for the todos that I added to this file